src/spectralanalysis.jl

```diff
@@ -67,29 +114,28 @@ More details can be found at
 If the given operator is non-hermitian a warning is given. This behavior
 can be turned off using the keyword `warning=false`.
 """
-function eigenenergies(op::DenseOpType, n::Int=length(basis(op)); warning=true)
-    b = basis(op)
+function eigenenergies(op::Operator, ds::LapackDiag; warning=true)
     if ishermitian(op)
-        D = eigvals(Hermitian(op.data), 1:n)
+        D = eigvals(Hermitian(op.data), 1:ds.n)
         return D
     else
         warning && @warn(nonhermitian_warning)
         D = eigvals(op.data)
         sort!(D, by=real)
-        return D[1:n]
+        return D[1:ds.n]
     end
 end
 
 """
 For sparse operators by default it only returns the 6 lowest eigenvalues.
 """
-eigenenergies(op::SparseOpType, n::Int=6; kwargs...) = eigenstates(op, n; kwargs...)[1]
-
-
-arithmetic_unary_error = QuantumOpticsBase.arithmetic_unary_error
-eigenstates(op::AbstractOperator, n::Int=0) = arithmetic_unary_error("eigenstates", op)
-eigenenergies(op::AbstractOperator, n::Int=0) = arithmetic_unary_error("eigenenergies", op)
+eigenenergies(op::Operator, ds::DiagStrategy; kwargs...) = eigenstates(op, ds; kwargs...)[1]
 
+function eigenenergies(op::AbstractOperator, n::Union{Int,Nothing}=nothing; kw...)
```

Comment:
This seems like the main entry to most of the code paths. Could you consider reorganizing the docstrings a bit, because now they seem to be at least partially outdated. A few options:

- move the information from most docstrings to only here and explain how the dispatch works. Then add very short docstrings to the other methods
- declare a `function eigenenergies end` without a method, and add the docstring to it.

Could you also add a brief sentence to the appropriate docstring explaining what is happening and that the library tries to detect the best choice on its own. And that the choice can be overwritten by using the appropriate type.